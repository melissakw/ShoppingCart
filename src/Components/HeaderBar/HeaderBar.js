import React from "react";
import { Link } from "react-router-dom";
import { useSelector } from "react-redux";
import ShoppingCartIcon from "@material-ui/icons/ShoppingCart";
import {
  Header,
  HomeLink,
  CartIcon,
  NumberOfItemsInCart
} from "../App.elements";

const HeaderBar = () => {
  const numOfItemsInCart = useSelector((state) =>
    state.product.cartItems.reduce((a, c) => a + +c.quantity, 0)
  );

  return (
    <Header>
      <HomeLink to="/">
        <h3>THE SHOP</h3>
      </HomeLink>
      <CartIcon>
        <Link to="/cart">
          <ShoppingCartIcon data-testid="shopping-icon" style={{ color: "black" }} />
        </Link>
        <NumberOfItemsInCart className="badge badge-dark">
          {numOfItemsInCart}
        </NumberOfItemsInCart>
      </CartIcon>
    </Header>
  );
};

export default HeaderBar;
