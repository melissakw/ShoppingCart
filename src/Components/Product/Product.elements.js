import styled from "styled-components";

export const ProductsContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const ProductCardContainer = styled.div`
  padding: 10px;
  width: 220px;
  height: 330px;
  margin: 10px;
  text-align: center;
  font-size: 16px;
`;

export const ProductText = styled.p`
  margin-top: "auto";
`;

export const ProductImage = styled.img`
  width: 150px;
  height: 200px;
`;

export const ProductDetails = styled.div`
  line-height: 30%;
`;

export const AddToCartButton = styled.button`
  background-color: black;
  color: white;
  border: none;
  padding: 10px;
  cursor: pointer;
  opacity: 0.6;
  transition: 0.4s;

  &:hover {
    opacity: 1;
  }
`;
