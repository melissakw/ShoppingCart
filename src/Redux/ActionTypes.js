export const ADD_ITEM = "ADD_ITEM";
export const REMOVE_ITEM = "REMOVE_ITEM";
export const EDIT_ITEM_QTY = "EDIT_ITEM_QTY";

export const ADD_TO_CART_LOADING = "ADD_TO_CART_LOADING";
export const ADD_TO_CART_FAILED = "ADD_TO_CART_FAILED";
