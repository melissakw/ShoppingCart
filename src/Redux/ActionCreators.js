import * as ActionTypes from "./ActionTypes";

export const postToCart = (product) => (dispatch) => {
  const productWithQty = {
    id: product.id,
    title: product.title,
    price: product.price,
    image: product.image,
    quantity: 1, //defualt quantity to 1 when adding item to cart
    stock: 10
  };
  dispatch(addToCart(productWithQty));
};

export const addToCart = (product) => ({
  type: ActionTypes.ADD_ITEM,
  payload: product
});

export const removeItem = (id) => ({
  type: ActionTypes.REMOVE_ITEM,
  payload: id
});

export const editItem = (product, editedQuantity) => ({
  type: ActionTypes.EDIT_ITEM_QTY,
  product: product,
  quantity: editedQuantity
});

export const addToCartLoading = () => ({
  type: ActionTypes.ADD_TO_CART_LOADING
});

export const addToCartFailed = (errmess) => ({
  type: ActionTypes.ADD_TO_CART_FAILED,
  payload: errmess
});
