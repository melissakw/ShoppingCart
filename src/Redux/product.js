import * as ActionTypes from "./ActionTypes";

export const product = (
  state = {
    isLoading: true,
    errMess: null,
    cartItems: []
  },
  action
) => {
  switch (action.type) {
    case ActionTypes.ADD_ITEM:
      let duplicateItem = false;
      let copyItems = state.cartItems.map((item) => {
        if (item.id === action.payload.id) {
          item.quantity = Math.min(item.quantity + 1, item.stock);
          duplicateItem = true;
        }
        return item;
      });

      if (duplicateItem) {
        return { ...state, cartItems: copyItems };
      } else {
        return {
          ...state,
          cartItems: [action.payload].concat(state.cartItems)
        };
      }

    case ActionTypes.REMOVE_ITEM:
      return {
        ...state,
        cartItems: state.cartItems.filter((item) => item.id !== action.payload)
      };
    case ActionTypes.EDIT_ITEM_QTY:
      let updatedProductWithNewQty = state.cartItems.map((item) => {
        if (item.id === action.product.id) {
          item.quantity = action.quantity;
        }
        return item;
      });
      return { ...state, cartItems: updatedProductWithNewQty };

    default:
      return state;
  }
};
