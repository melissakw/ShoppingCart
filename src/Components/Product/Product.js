import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { postToCart } from "../../Redux/ActionCreators";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import {
  ProductsContainer,
  ProductCardContainer,
  ProductText,
  ProductImage,
  ProductDetails,
  AddToCartButton
} from "./Product.elements";

const Product = ({ products }) => {
  const dispatch = useDispatch();
  const [addingToCart, setAddingToCart] = useState(null);
  const noStyle = { boxShadow: "0px 0px 0px white" };
  const centerItems = { justifyContent: "center" };

  return (
    <ProductsContainer>
      {products.map((prod) => {
        return (
          <Card key={prod.id} style={noStyle}>
            <ProductCardContainer  className="d-flex flex-column">
              <ProductImage src={prod.image} alt={prod.title} />

              <ProductDetails>
                <div>
                  <CardContent>
                    <ProductText>
                      {prod.title.slice(0, 10)}
                    </ProductText>
                    <ProductText>
                      ${prod.price}
                    </ProductText>
                  </CardContent>
                </div>

                <CardActions style={centerItems}>
                  <AddToCartButton
                    data-testid="add-to-cart"
                    onClick={() => {
                      setAddingToCart(prod.id);
                      dispatch(postToCart(prod));
                      setTimeout(() => setAddingToCart(null), 1800);
                    }}
                  >
                    {addingToCart === prod.id ? "Added to cart" : "Add to cart"}
                  </AddToCartButton>
                </CardActions>
              </ProductDetails>
            </ProductCardContainer>
          </Card>
        );
      })}
    </ProductsContainer>
  );
};

export default Product;
