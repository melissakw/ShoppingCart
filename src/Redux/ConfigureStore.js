import { createStore, applyMiddleware } from "redux";
import { persistStore, persistCombineReducers } from "redux-persist";
import storage from "redux-persist/lib/storage";
import thunk from "redux-thunk";
import logger from "redux-logger";

import { product } from "./product";

export const ConfigureStore = () => {
  const config = {
    key: "root",
    storage: storage,
    debug: true
  };

  const store = createStore(
    persistCombineReducers(config, {
      product
    }),
    applyMiddleware(thunk, logger)
  );

  const persistor = persistStore(store);
  return { persistor, store };
};
