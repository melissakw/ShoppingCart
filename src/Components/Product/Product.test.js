import React from 'react';
import Product from './Product';
import HeaderBar from '../HeaderBar/HeaderBar';
import { useSelector, useDispatch } from 'react-redux'; 
import {render, screen, fireEvent} from "@testing-library/react";
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from "react-redux";
import { ConfigureStore } from "../../Redux/ConfigureStore";
import { BrowserRouter, Switch, Route, Link } from 'react-router-dom';


test("renders Product correctly", () => {
	const { store } = ConfigureStore();
	const items = [{"id":1,"title":"Backpack","price":109.95,"description":"some description","category":"clothing","image":"https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"},
  {"id":2,"title":"shirt ","price":22.3,"description":"some description","category":"clothing","image":"https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"}];
  
	const tree = renderer.create(
		<Provider store={store}>
			<BrowserRouter>
				<Product products={items}/>
			</BrowserRouter>
	/ </Provider>
	).toJSON();
  expect(tree).toMatchSnapshot();
});

// test('calls onClick when "Add To Cart" is clicked', () => {
//   const mockedPostToCart = jest.fn();
//   const { getByTestId } = render(<button onClick={mockedPostToCart} />);

// 	fireEvent.click(getByTestId(/add-to-cart/i));
// 	expect(mockedPostToCart).toHaveBeenCalled();
// 	});
