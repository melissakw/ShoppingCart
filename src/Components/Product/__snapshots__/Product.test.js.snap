// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Product correctly 1`] = `
Array [
  <div
    className="sc-bdfBwQ pdzVR"
  >
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      style={
        Object {
          "boxShadow": "0px 0px 0px white",
        }
      }
    >
      <div
        className="sc-gsTCUz jsccXy d-flex flex-column"
      >
        <img
          alt="Backpack"
          className="sc-hKgILt Ibfk"
          src="https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg"
        />
        <div
          className="sc-eCssSg jrJvnJ"
        >
          <div>
            <div
              className="MuiCardContent-root"
            >
              <p
                className="sc-dlfnbm ibNFrJ"
              >
                Backpack
              </p>
              <p
                className="sc-dlfnbm ibNFrJ"
              >
                $
                109.95
              </p>
            </div>
          </div>
          <div
            className="MuiCardActions-root MuiCardActions-spacing"
            style={
              Object {
                "justifyContent": "center",
              }
            }
          >
            <button
              className="sc-jSgupP iYGaRU"
              data-testid="add-to-cart"
              onClick={[Function]}
            >
              Add to cart
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
      style={
        Object {
          "boxShadow": "0px 0px 0px white",
        }
      }
    >
      <div
        className="sc-gsTCUz jsccXy d-flex flex-column"
      >
        <img
          alt="shirt "
          className="sc-hKgILt Ibfk"
          src="https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg"
        />
        <div
          className="sc-eCssSg jrJvnJ"
        >
          <div>
            <div
              className="MuiCardContent-root"
            >
              <p
                className="sc-dlfnbm ibNFrJ"
              >
                shirt 
              </p>
              <p
                className="sc-dlfnbm ibNFrJ"
              >
                $
                22.3
              </p>
            </div>
          </div>
          <div
            className="MuiCardActions-root MuiCardActions-spacing"
            style={
              Object {
                "justifyContent": "center",
              }
            }
          >
            <button
              className="sc-jSgupP iYGaRU"
              data-testid="add-to-cart"
              onClick={[Function]}
            >
              Add to cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "/ ",
]
`;
